grpc/pb-marshalling.go

```diff
@@ -0,0 +1,263 @@
+var ErrNilReturn = CodedError(codes.Internal, "unexpected nil return value with no error")
+
+// This file defines functions to translate between the protobuf types and the
+// code types.
+
+func authzMetaToPB(authz core.Authorization) (*vapb.AuthzMeta, error) {
+	return &vapb.AuthzMeta{
+		Id:    &authz.ID,
+		RegID: &authz.RegistrationID,
+	}, nil
+}
+
+func pbToAuthzMeta(in *vapb.AuthzMeta) (core.Authorization, error) {
+	if in == nil || in.Id == nil || in.RegID == nil {
+		return core.Authorization{}, ErrMissingParameters
+	}
+	return core.Authorization{
+		ID:             *in.Id,
+		RegistrationID: *in.RegID,
+	}, nil
+}
+
+func jwkToString(jwk *jose.JsonWebKey) (string, error) {
+	bytes, err := jwk.MarshalJSON()
+	return string(bytes), err
+}
+
+func stringToJWK(in string) (*jose.JsonWebKey, error) {
+	var jwk = new(jose.JsonWebKey)
+	err := jwk.UnmarshalJSON([]byte(in))
+	if err != nil {
+		return nil, err
+	}
+	return jwk, nil
+}
+
+func problemDetailsToPB(prob *probs.ProblemDetails) (*corepb.ProblemDetails, error) {
+	if prob == nil {
+		// nil problemDetails is valid
+		return nil, nil
+	}
+	pt := string(prob.Type)
+	st := int32(prob.HTTPStatus)
+	return &corepb.ProblemDetails{
+		ProblemType: &pt,
+		Detail:      &prob.Detail,
+		HttpStatus:  &st,
+	}, nil
+}
+
+func pbToProblemDetails(in *corepb.ProblemDetails) (*probs.ProblemDetails, error) {
+	if in == nil {
+		// nil problemDetails is valid
+		return nil, nil
+	}
+	if in.ProblemType == nil || in.Detail == nil {
+		return nil, ErrMissingParameters
+	}
+	prob := &probs.ProblemDetails{
+		Type:   probs.ProblemType(*in.ProblemType),
+		Detail: *in.Detail,
+	}
+	if in.HttpStatus != nil {
+		prob.HTTPStatus = int(*in.HttpStatus)
+	}
+	return prob, nil
+}
+
+func vaChallengeToPB(challenge core.Challenge) (*corepb.Challenge, error) {
+	accountKey, err := jwkToString(challenge.AccountKey)
+	if err != nil {
+		return nil, err
+	}
+	st := string(challenge.Status)
+	return &corepb.Challenge{
+		Id:               &challenge.ID,
+		Type:             &challenge.Type,
+		Status:           &st,
+		Token:            &challenge.Token,
+		AccountKey:       &accountKey,
+		KeyAuthorization: &challenge.ProvidedKeyAuthorization,
+	}, nil
+}
+
+func pbToVAChallenge(in *corepb.Challenge) (challenge core.Challenge, err error) {
+	if in == nil {
+		return core.Challenge{}, ErrMissingParameters
+	}
+	if in.AccountKey == nil || in.Id == nil || in.Type == nil || in.Status == nil || in.Token == nil || in.KeyAuthorization == nil {
+		return core.Challenge{}, ErrMissingParameters
+	}
+	jwk, err := stringToJWK(*in.AccountKey)
+	if err != nil {
+		return
+	}
+	return core.Challenge{
+		ID:                       *in.Id,
+		Type:                     *in.Type,
+		Status:                   core.AcmeStatus(*in.Status),
+		Token:                    *in.Token,
+		AccountKey:               jwk,
+		ProvidedKeyAuthorization: *in.KeyAuthorization,
+	}, nil
+}
+
+func ipAddrToString(ip net.IP) (string, error) {
```

Comment:
What does this function and stirngToIPAddr add? It seems like they are trivial wrappers around `MarshalText` and `UnmarshalText`?